src/core/lwt_pool.mli

```diff
@@ -27,31 +27,45 @@
     The pool also provides a limit on the number of connections that can
     simultaneously be open.
 
-    Note that pools are not for keeping Lwt threads. Lwt threads are very cheap
-    to create and are pure. It is neither desirable nor possible to reuse them.
-    If you want to have a pool of {e system} threads, consider module
+    Note that pools are not for keeping Lwt promises. Lwt promises are very
+    cheap to create. It is neither desirable nor possible to reuse them.  If
+    you want to have a pool of {e system} threads, consider using
     [Lwt_preemptive]. *)
 
 type 'a t
-  (** Pools containing values of type ['a]. *)
+  (** Pools containing elements of type ['a]. *)
 
 val create :
   int ->
   ?check : ('a -> (bool -> unit) -> unit) ->
   ?validate : ('a -> bool Lwt.t) ->
+  ?dispose : ('a -> unit Lwt.t) ->
   (unit -> 'a Lwt.t) -> 'a t
-  (** [create n ?check ?validate f] creates a new pool with at most
-      [n] elements. [f] is the function to use to create a new element
+  (** [create n ?check ?validate ?dispose f] creates a new pool with at most
+      [n] elements. [f] is the function to use to create a new element.
       Elements are created on demand.
 
-      An element of the pool is validated by the optional [validate]
-      function before its {!use}. Invalid elements are re-created.
+      An element of the pool is validated by the optional [validate] function
+      before it is accessed by {!use}. Invalid elements are passed to [dispose]
+      and then re-created with [f].
 
-      The optional function [check] is called after a [use] of an
-      element failed. It must call its argument exactly once with
-      [true] if the element is still valid and [false] otherwise. *)
+      If a call to {!use} fails with a pool element that element will be passed
+      to the optional function [check] as [check element callback].  [check]
+      must call [callback] exactly once with [true] if [element] is still valid
+      and [false] otherwise.  If [check] calls [callback false] then [dispose]
+      will be run on [element].
```

Comment:
I'm a bit confused on when `dispose` is called. At a glance, it seems that both `validate` and `check` will call `dispose`. Will `dispose` be called twice? If not, how is the order of calling done?

Also this:

> If a call to {!use} fails with a pool element

As my comment in the issue, I think it's unclear why a `use` would fail. Does "fail" equals "trying to use an invalid pool element" (due to `validate`)?

Reply:
I believe "fail" means raising an exception or rejecting the `'b Lwt.t` promise that `use`'s callback returns. It should definitely be clarified in the comment, because I also am guessing about what it means.

Reply:
Ah, that actually makes sense. I think it's the whole `validate` and `invalid elements` on `use` paragraph just above it that confused me, I thought they are still talking about the same (or at least related) thing.

Reply:
I tried to tweak the wording at bit - does this make the use of each argument any clearer?  Suggestions are welcome - the wording has been tricky to get right.

Reply:
The new docs are pretty great!